### PR TITLE
[FIX] payment_stripe: use idempotency key only when offline

### PR DESCRIPTION
**Pull request by abla001**
To Reproduce
============
- on a db with sales, website and Stripe payment aquirer activated
- on the portal of a user add two payment methods, an expired card
and a valid one
- create a SO and generate its payment link
- try to pay with the invalid card first -> and error (card was declined)
- try to pay with the valid card -> error (same idempotency key)

Problem
=======
the generation of the idempotency key is based only on dbuuid, transaction's
reference and the scope. So in this use case it will generate the same key.

Solution
========
The idempotency key prevents issues where the customer is charged twice for
the same thing. In this case, we don't want to prevent anything since the
customer is on the page. It's suggested to use idempotency keys only for offline
payment when the customer is not in front of the payment page (e.g. when the cron
charges the customer for his subscription)

opw-3091354

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
doesn't seem too bad (since we only use this one time), but in stable it might cleaner to have a method that give a "payment method" overridden by stripe (to `self.token_id.stripe_payment_method`) , and always call `self.get_payment_method()` in `generate_idempotency_key`.

**Comment by nle-odoo** [comment]
> `rd-security` can you please check the security of this PR, is there a possibility to inject something using payment_method argument ?

I think the normal flow is: when all is set for this to be merged: ask security team review and explain why there is no security issue in your code, for example:

use sha1 was present before the commit (approved in https://github.com/odoo/odoo/pull/101243#issuecomment-1269559196 with reason "SHA1 : should not collide, uuid + ref (unique per db)") and we only add a parameter that is coming from stripe and is an unique token of alphabetic and numeric character that is unique for a payment method and can't be set by the customer.

File: addons/payment_stripe/models/payment_transaction.py
Comment (chevalierv) [inline addons/payment_stripe/models/payment_transaction.py:203]: We don't want to keep users from retrying a payment. In this case, the payment method differs but there could be other possibilities (like a customer that does not have enough money in his bank account. He will get an "Insufficient funds" error, then make an instant transfer and retry. It won't work since the payment method is the same).

The idempotency key prevents issues where the customer is charged twice for the same thing. In this case, we don't want to prevent anything since the customer is on the page. I discussed this matter with the payment team and we suggest using idempotency keys only for offline payment when the customer is not in front of the payment page (e.g. when the cron charges the customer for his subscription).
```suggestion
            idempotency_key=payment_utils.generate_idempotency_key(
                self, scope='payment_intents_token'
            ) if self.operation == 'offline' else None,
```

**Review by chevalierv — CHANGES_REQUESTED** [review]
Hello @abla001,

Thanks for your work on this opw!

I don't think we need to change the `generate_idempotency_key` method. I've left a comment explaining why with a suggestion to fix the issue. 

Don't hesitate to come back to me if you have any questions.

Have a good evening :night_with_stars:

**Comment by abla001** [comment]
Hello @chevalierv,
thank you for the feedback :pray: ,
I updated the PR with the suggested changes

**Review by chevalierv — APPROVED** [review]
LGTM :+1:

**Review by AntoineVDV — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**
